PR: https://github.com/odoo/odoo/pull/178007

From: 4dfa42df501d6b7bca4586a40bc0388d9f3b118b
From: Fabien Pinckaers
Date: 2024-09-09 14:37:55

Breaking data model changes scores: del:14 + add:2, change matches:
+    mailing_ids = fields.One2many('mailing.mailing', 'card_campaign_id')
+    card_ids = fields.One2many('card.card', inverse_name='campaign_id')
-    card_element_ids = fields.One2many('card.campaign.element', inverse_name='campaign_id', copy=True)
-    preview_record_url = fields.Char('Preview Record Link', compute="_compute_preview_record_url")
-    res_model = fields.Selection(related="campaign_id.res_model", precompute=True, readonly=True, store=True)
-    campaign_id = fields.Many2one('card.campaign', required=True, ondelete='cascade')
-    card_element_role = fields.Selection([
-    card_element_image = fields.Image(attachment=False, compute="_compute_value_fields", readonly=False, store=True)
-    card_element_text = fields.Text(compute="_compute_value_fields", readonly=False, store=True)
-    card_element_text_value = fields.Char(string="Text", compute="_compute_card_element_text_value")
-    field_path = fields.Char(compute="_compute_value_fields", readonly=False, store=True)
-    text_color = fields.Char(compute="_compute_value_fields", readonly=False, store=True)
-    render_type = fields.Selection([('image', 'Image'), ('text', 'User Text')], default='text', required=True)
-    value_type = fields.Selection([('static', 'Manual'), ('field', 'Dynamic')], default='static', required=True)
-    record_ref = fields.Reference(string='Record', selection='_selection_record_ref', compute='_compute_record_ref')

Total Changes: 603

[IMP] marketing_cards

remove card_campaign_elements, and card_element_ids: use static fields
on the campaign instead: allows a better UX and much less code.

compute images directly, when creating a card (on sending email) to
avoid having the public user to render images in "sudo()". That allows
also to remove the whitelist mechanism as there is no more sudo(): the
user builds image, with his own rights.

improved views: list, form, etc.

Added a customized title and message to motivate sharing.

Simplified code for Preview: previews are a card like all others, but
active=False. Preview is now a button, instead of a URL field.

fF we pass the card_id in the URL (instead of res_id), there is no need
for a hash_token; by definition, all cards are meant to be published.
That makes URLs less ugly. It also allow to remove all sudo() as cards
can be read by anyone.

Improved CSS of card layouts (e.g. avoid messing up if text too lang),
replace better image, compress better some images.

Temporarily removing tests; too specific to current implementation to
be rewritten (no need for a separate module, based on demo data)

Part-of: odoo/odoo#178007
Related: odoo/enterprise#68957
Related: odoo/upgrade#6420
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/marketing_card/models/__init__.py
+++ b/addons/marketing_card/models/__init__.py
@@ -1,7 +1,6 @@
 from . import card_campaign
-from . import card_campaign_element
 from . import card_campaign_tag
 from . import card_template
 from . import card_card
-from . import models
+from . import mailing_mailing
 from . import utm_source

--- a/addons/marketing_card/models/card_campaign.py
+++ b/addons/marketing_card/models/card_campaign.py
@@ -1,9 +1,6 @@
 import base64
-from lxml import html
-from itertools import count
 
-from odoo import _, api, Command, fields, models
-from odoo.tools.misc import hmac
+from odoo import _, api, fields, models, exceptions
 
 from .card_template import TEMPLATE_DIMENSIONS
 
@@ -17,22 +14,6 @@ class CardCampaign(models.Model):
     def _default_card_template_id(self):
         return self.env['card.template'].search([], limit=1)
 
-    def default_get(self, fields_list):
-        default_vals = super().default_get(fields_list)
-        if 'card_element_ids' in fields_list and 'card_element_ids' not in default_vals:
-            default_vals.setdefault('card_element_ids', [
-                Command.create({'card_element_role': 'background', 'render_type': 'image'}),
-                Command.create({'card_element_role': 'header', 'render_type': 'text'}),
-                Command.create({'card_element_role': 'subheader', 'render_type': 'text'}),
-                Command.create({'card_element_role': 'section_1', 'render_type': 'text'}),
-                Command.create({'card_element_role': 'subsection_1', 'render_type': 'text'}),
-                Command.create({'card_element_role': 'subsection_2', 'render_type': 'text'}),
-                Command.create({'card_element_role': 'button', 'render_type': 'text'}),
-                Command.create({'card_element_role': 'image_1', 'render_type': 'image'}),
-                Command.create({'card_element_role': 'image_2', 'render_type': 'image'}),
-            ])
-        return default_vals
-
     def _get_model_selection(self):
         """Hardcoded list of models, checked against actually-present models."""
         allowed_models = ['res.partner', 'event.track', 'event.booth', 'event.registration']
@@ -47,25 +28,63 @@ class CardCampaign(models.Model):
     card_click_count = fields.Integer(compute='_compute_card_stats')
     card_share_count = fields.Integer(compute='_compute_card_stats')
 
+    mailing_ids = fields.One2many('mailing.mailing', 'card_campaign_id')
+    mailing_count = fields.Integer(compute='_compute_mailing_count')
+
+    card_ids = fields.One2many('card.card', inverse_name='campaign_id')
     card_template_id = fields.Many2one('card.template', string="Design", default=_default_card_template_id, required=True)
-    card_element_ids = fields.One2many('card.campaign.element', inverse_name='campaign_id', copy=True)
-    image_preview = fields.Image(compute='_compute_image_preview', readonly=True, store=True, compute_sudo=False, attachment=False)
+    image_preview = fields.Image(compute='_compute_image_preview', compute_sudo=False, readonly=True, store=True, attachment=False)
     link_tracker_id = fields.Many2one('link.tracker', ondelete="restrict")
-    res_model = fields.Selection(string="Model Name", selection=_get_model_selection,
-                                 compute="_compute_res_model", copy=True, precompute=True,
-                                 readonly=False, required=True, store=True)
+    res_model = fields.Selection(
+        string="Model Name", compute='_compute_res_model', selection='_get_model_selection',
+        precompute=True, readonly=True, required=True, store=True,
+    )
 
     post_suggestion = fields.Text(help="Description below the card and default text when sharing on X")
-    preview_record_ref = fields.Reference(string="Preview Record", selection="_selection_preview_record_ref")
-    preview_record_url = fields.Char('Preview Record Link', compute="_compute_preview_record_url")
-    reward_message = fields.Html(string='Thanks to You Message')
-    reward_target_url = fields.Char(string='Reward Link')
+    preview_record_ref = fields.Reference(string="Preview On", selection="_get_model_selection", required=True)
     tag_ids = fields.Many2many('card.campaign.tag', string='Tags')
-    target_url = fields.Char(string='Shared Link')
+    target_url = fields.Char(string='Post Link')
     target_url_click_count = fields.Integer(related="link_tracker_id.count")
 
     user_id = fields.Many2one('res.users', string='Responsible', default=lambda self: self.env.user, domain="[('share', '=', False)]")
 
+    reward_message = fields.Html(string='Thanks to You Message')
+    reward_target_url = fields.Char(string='Reward Link')
+    request_title = fields.Char('Request', default=lambda self: _('Help us share the news'))
+    request_description = fields.Text('Request Description')
+
+    # Static Content fields
+    content_background = fields.Image('Background')
+    content_button = fields.Char('Button')
+
+    # Dynamic Content fields
+    content_header = fields.Char('Header')
+    content_header_dyn = fields.Boolean('Is Dynamic Header')
+    content_header_path = fields.Char('Header Path')
+    content_header_color = fields.Char('Header Color')
+
+    content_sub_header = fields.Char('Sub-Header')
+    content_sub_header_dyn = fields.Boolean('Is Dynamic Sub-Header')
+    content_sub_header_path = fields.Char('Sub-Header Path')
+    content_sub_header_color = fields.Char('Sub Header Color')
+
+    content_section = fields.Char('Section')
+    content_section_dyn = fields.Boolean('Is Dynamic Section')
+    content_section_path = fields.Char('Section Path')
+
+    content_sub_section1 = fields.Char('Sub-Section 1')
+    content_sub_section1_dyn = fields.Boolean('Is Dynamic Sub-Section 1')
+    content_sub_section1_path = fields.Char('Sub-Section 1 Path')
+
+    content_sub_section2 = fields.Char('Sub-Section 2')
+    content_sub_section2_dyn = fields.Boolean('Is Dynamic Sub-Section 2')
+    content_sub_section2_path = fields.Char('Sub-Section 2 Path')
+
+    # images are always dynamic
+    content_image1_path = fields.Char('Dynamic Image 1')
+    content_image2_path = fields.Char('Dynamic Image 2')
+
+    @api.depends('card_ids')
     def _compute_card_stats(self):
         cards_by_status_count = self.env['card.card']._read_group(
             domain=[('campaign_id', 'in', self.ids)],
@@ -82,47 +101,43 @@ class CardCampaign(models.Model):
             # shared cards are implicitly visited
             if status == 'shared':
                 campaign.card_share_count += card_count
-            if status:
+            if status in ('shared', 'visited'):
                 campaign.card_click_count += card_count
             campaign.card_count += card_count
 
-    @api.depends('body_html', 'card_element_ids', 'preview_record_ref', 'res_model', 'card_element_ids.card_element_role',
-                 'card_element_ids.card_element_image', 'card_element_ids.card_element_text', 'card_element_ids.field_path',
-                 'card_element_ids.text_color', 'card_element_ids.render_type', 'card_element_ids.value_type')
+    @api.model
+    def _get_render_fields(self):
+        return [
+            'body_html', 'content_background', 'content_image1_path', 'content_image2_path', 'content_button', 'content_header',
+            'content_header_dyn', 'content_header_path', 'content_header_color', 'content_sub_header',
+            'content_sub_header_dyn', 'content_sub_header_path', 'content_section', 'content_section_dyn',
+            'content_section_path', 'content_sub_section1', 'content_sub_section1_dyn', 'content_sub_header_color',
+            'content_sub_section1_path', 'content_sub_section2', 'content_sub_section2_dyn', 'content_sub_section2_path'
+        ]
+
+    @api.depends(lambda self: self._get_render_fields() + ['preview_record_ref'])
     def _compute_image_preview(self):
-        rendered_campaigns = self.filtered('card_template_id.body').filtered('card_element_ids')
-        (self - rendered_campaigns).image_preview = False
-
-        for campaign in rendered_campaigns:
+        for campaign in self:
             if campaign.preview_record_ref and campaign.preview_record_ref.exists():
                 image = campaign._get_image_b64(campaign.preview_record_ref)
             else:
-                image = campaign._get_generic_image_b64()
-            if image is not None:
-                campaign.image_preview = image
-
-    @api.depends('preview_record_ref')
-    def _compute_preview_record_url(self):
-        self.preview_record_url = False
-        for campaign in self.filtered('preview_record_ref'):
-            if campaign._origin.id:
-                campaign.preview_record_url = campaign._get_preview_url_from_res_id(campaign.preview_record_ref.id)
+                image = False
+            campaign.image_preview = image
+
+    @api.depends('mailing_ids')
+    def _compute_mailing_count(self):
+        self.mailing_count = 0
+        mailing_counts = self.env['mailing.mailing']._read_group(
+            [('card_campaign_id', 'in', self.ids)], ['card_campaign_id'], ['__count']
+        )
+        for campaign, mailing_count in mailing_counts:
+            campaign.mailing_count = mailing_count
 
     @api.depends('preview_record_ref')
     def _compute_res_model(self):
         for campaign in self:
-            if campaign.preview_record_ref:
-                campaign.res_model = campaign.preview_record_ref._name
-            elif not campaign.res_model:
-                campaign.res_model = 'res.partner'
-
-    @api.onchange('res_model', 'preview_record_ref')
-    def _onchange_res_model(self):
-        for campaign in self:
-            if not campaign._origin.res_model:
-                continue
-            if campaign._origin.res_model != campaign.res_model:
-                campaign.card_element_ids.value_type = 'static'
+            preview_model = campaign.preview_record_ref and campaign.preview_record_ref._name
+            campaign.res_model = preview_model or campaign.res_model or 'res.partner'
 
     @api.model_create_multi
     def create(self, create_vals):
@@ -143,17 +158,30 @@ class CardCampaign(models.Model):
 
     def write(self, vals):
         link_tracker_vals = {}
-
+        if vals.keys() & set(self._get_render_fields()):
+            self.env['card.card'].search([('campaign_id', 'in', self.ids)]).requires_sync = True
         if 'target_url' in vals:
             link_tracker_vals['url'] = vals['target_url'] or self.env['card.campaign'].get_base_url()
         if link_tracker_vals:
             self.link_tracker_id.sudo().write(link_tracker_vals)
 
-        return super().write(vals)
+        # write and detect model changes on actively-used campaigns
+        original_models = self.mapped('res_model')
 
-    @api.model
-    def _selection_preview_record_ref(self):
-        return self._fields['res_model']._description_selection(self.env)
+        write_res = super().write(vals)
+
+        updated_model_campaigns = self.env['card.campaign'].browse([
+            campaign.id for campaign, new_model, old_model
+            in zip(self, self.mapped('res_model'), original_models)
+            if new_model != old_model
+        ])
+        for campaign in updated_model_campaigns:
+            if campaign.card_count:
+                raise exceptions.ValidationError(_(
+                    "Model of campaign %(campaign)s may not be changed as it already has cards",
+                    campaign=campaign.display_name,
+                ))
+        return write_res
 
     def action_view_cards(self):
         self.ensure_one()
@@ -176,6 +204,34 @@ class CardCampaign(models.Model):
             'domain': [('campaign_id', '=', self.id)],
         }
 
+    def action_view_mailings(self):
+        self.ensure_one()
+        return {
+            'name': _('%(card_campaign_name)s Mailings', card_campaign_name=self.name),
+            'type': 'ir.actions.act_window',
+            'res_model': 'mailing.mailing',
+            'domain': [('card_campaign_id', '=', self.id)],
+            'view_mode': 'list,form',
+            'target': 'current',
+        }
+
+    def action_preview(self):
+        self.ensure_one()
+        card = self.env['card.card'].with_context(active_test=False).search([
+            ('campaign_id', '=', self.id),
+            ('res_id', '=', self.preview_record_ref.id),
+        ])
+        if card:
+            card.image = self.image_preview
+        else:
+            card = self.env['card.card'].create({
+                'campaign_id': self.id,
+                'res_id': self.preview_record_ref.id,
+                'image': self.image_preview,
+                'active': False,
+            })
+        return {'type': 'ir.actions.act_url', 'url': card._get_path('preview'), 'target': 'new'}
+
     def action_share(self):
         self.ensure_one()
         return {
@@ -184,6 +240,8 @@ class CardCampaign(models.Model):
             'res_model': 'mailing.mailing',
             'context': {
                 'default_subject': self.name,
+                'default_card_campaign_id': self.id,
+                'default_mailing_model_id': self.env['ir.model']._get_id(self.res_model),
                 'default_body_arch': f"""
 <div class="o_layout oe_unremovable oe_unmovable bg-200 o_empty_theme" data-name="Mailing">
 <style id="design-element"></style>
@@ -191,6 +249,15 @@ class CardCampaign(models.Model):
 <div class="row">
 <div class="col o_mail_no_options o_mail_wrapper_td bg-white oe_structure o_editable theme_selection_done">
 
+<div class="s_text_block o_mail_snippet_general pt24 pb24" style="padding-left: 15px; padding-right: 15px;" data-snippet="s_text_block" data-name="Text">
+    <div class="container s_allow_columns">
+        <p class="o_default_snippet_text">Hello everyone</p>
+        <p class="o_default_snippet_text">Here's the link to advertise your participation.
+        <br> Your help with this promotion would be greatly appreciated!`</p>
+        <p class="o_default_snippet_text">Many thanks</p>
+    </div>
+</div>
+
 <div class="s_call_to_share_card o_mail_snippet_general" style="padding-top: 10px; padding-bottom: 10px;">
     <table width="100%" border="0" cellspacing="0" cellpadding="0">
         <tbody>
@@ -220,62 +287,58 @@ class CardCampaign(models.Model):
         if not self.card_template_id.body:
             return ''
 
-        # check this early to get a better error message
-        for element in self.card_element_ids:
-            if element._origin.field_path != element.field_path:
-                element._check_fields()
-
         image_bytes = self.env['ir.actions.report']._run_wkhtmltoimage(
             [self._render_field('body_html', record.ids, add_context={'card_campaign': self})[record.id]],
             *TEMPLATE_DIMENSIONS
         )[0]
         return image_bytes and base64.b64encode(image_bytes)
 
-    def _get_generic_image_b64(self):
-        """Render a single preview image with no record."""
-        rendered_body = self.env['ir.qweb']._render(
-            html.fromstring(self.body_html),
-            self._render_eval_context() | {
-                'card_campaign': self,
-                'preview_values': {
-                    'header': _('Title'),
-                    'subheader': _('Subtitle'),
-                }
-            },
-            raise_on_code=False,
-        )
-        image_bytes = self.env['ir.actions.report']._run_wkhtmltoimage(
-            [rendered_body],
-            *TEMPLATE_DIMENSIONS
-        )[0]
-        return image_bytes and base64.b64encode(image_bytes)
-
     # ==========================================================================
     # Card creation
     # ==========================================================================
 
-    def _generate_card_hash_token(self, record_id):
-        """Generate a token for a specific recipient of this campaign."""
-        self.ensure_one()
-        return hmac(self.env(su=True), ('marketing_card', self._name, self._origin.id, self.name), record_id)
-
-    def _get_or_create_cards_from_res_ids(self, res_ids):
-        """Create missing cards for the given ids."""
+    def _update_cards(self, domain, auto_commit=False):
+        """Create missing cards and update cards if necessary based for the domain."""
         self.ensure_one()
-        cards = self.env['card.card'].search_fetch([('campaign_id', '=', self.id), ('res_id', 'in', res_ids)], ['res_id'])
-        missing_ids = set(res_ids) - set(cards.mapped('res_id'))
-        cards += self.env['card.card'].create([{'campaign_id': self.id, 'res_id': missing_id} for missing_id in missing_ids])
-
-        # order based on input
-        res_order = dict(zip(res_ids, count()))
-        return cards.sorted(key=lambda card: res_order[card.res_id])
-
-    def _get_preview_url_from_res_id(self, res_id):
-        return self._get_card_path(res_id, 'preview')
+        TargetModel = self.env[self.res_model]
+        res_ids = TargetModel.search(domain).ids
+        cards = self.env['card.card'].with_context(active_test=False).search_fetch([
+            ('campaign_id', '=', self.id),
+            ('res_id', 'in', res_ids),
+        ], ['res_id', 'requires_sync'])
+        # update active and res_model for preview cards
+        cards.active = True
+        self.env['card.card'].create([
+            {'campaign_id': self.id, 'res_id': res_id}
+            for res_id in set(res_ids) - set(cards.mapped('res_id'))
+        ])
 
-    def _get_card_path(self, res_id, suffix):
-        self.ensure_one()
-        return f'{self.get_base_url()}/cards/{self._origin.id}/{res_id}/{self._generate_card_hash_token(res_id)}/{suffix}'
+        # render by batch of 100 to avoid losing progress in case of time out
+        updated_cards = self.env['card.card']
+        while cards := self.env['card.card'].search_fetch([
+            ('requires_sync', '=', True),
+            ('campaign_id', '=', self.id),
+            ('res_id', 'in', res_ids),
+        ], ['res_id'], limit=100):
+            # no need to autocommit if it can be done in one batch
+            if auto_commit and updated_cards:
+                self.env.cr.commit()
+                # avoid keeping hundreds of jpegs in memory
+                self.env['card.card'].invalidate_model(['image'])
+            TargetModelPrefetch = TargetModel.with_prefetch(cards.mapped('res_id'))
+            for card in cards.filtered('requires_sync'):
+                card.write({
+                    'image': self._get_image_b64(TargetModelPrefetch.browse(card.res_id)),
+                    'requires_sync': False,
+                    'active': True,
+                })
+            cards.flush_recordset()
+            updated_cards += cards
+        return updated_cards
+
+    def _get_url_from_res_id(self, res_id, suffix='preview'):
+        card = self.env['card.card'].search([('campaign_id', '=', self.id), ('res_id', '=', res_id)])
+        return card and card._get_path(suffix) or self.target_url
 
     # ==========================================================================
     # Mail render mixin / Render utils
@@ -287,23 +350,28 @@ class CardCampaign(models.Model):
         for campaign in self:
             campaign.render_model = campaign.res_model
 
-    def _get_card_element_values(self, record, preview_values):
-        """Helper to get the right value for each element when rendering."""
+    def _get_card_element_values(self, record):
+        """Helper to get the right value for dynamic fields."""
         self.ensure_one()
-        value_from_role = {}
-        default_values = {
-            'background': self.card_template_id.default_background
+        result = {
+            'image1': images[0] if (images := self.content_image1_path and record.mapped(self.content_image1_path)) else False,
+            'image2': images[0] if (images := self.content_image2_path and record.mapped(self.content_image2_path)) else False,
         }
-        for element in self.card_element_ids:
-            value = element._get_render_value(record)
-            if not value and element.card_element_role in default_values:
-                value = default_values[element.card_element_role]
-            if not value and preview_values and element.card_element_role in preview_values and element.value_type == 'static':
-                value = preview_values[element.card_element_role]
-            if not value and not record:
-                value = element._get_placeholder_value()
-            value_from_role[element.card_element_role] = value
-
-        # in qweb t-out of "False" effectively removed the element while '' does not
-        # we force everything to '' to be consistent
-        return {element: val or '' for element, val in value_from_role.items()}
+        campaign_text_element_fields = (
+            ('header', 'content_header', 'content_header_dyn', 'content_header_path'),
+            ('sub_header', 'content_sub_header', 'content_sub_header_dyn', 'content_sub_header_path'),
+            ('section', 'content_section', 'content_section_dyn', 'content_section_path'),
+            ('sub_section1', 'content_sub_section1', 'content_sub_section1_dyn', 'content_sub_section1_path'),
+            ('sub_section2', 'content_sub_section2', 'content_sub_section2_dyn', 'content_sub_section2_path'),
+        )
+        for el, text_field, dyn_field, path_field in campaign_text_element_fields:
+            if not self[dyn_field]:
+                result[el] = self[text_field]
+            else:
+                try:
+                    m = record.mapped(self[path_field])
+                    result[el] = m and m[0] or False
+                except (AttributeError, KeyError):
+                    # for generic image, or if field incorrect, return name of field
+                    result[el] = self[path_field]
+        return result

--- a/addons/marketing_card/models/card_campaign_element.py
+++ b/None
@@ -1,116 +0,0 @@
-import base64
-
-from odoo import _, api, exceptions, fields, models
-
-
-class CardCampaignElement(models.Model):
-    _name = 'card.campaign.element'
-    _description = 'Marketing Card Campaign Element'
-
-    res_model = fields.Selection(related="campaign_id.res_model", precompute=True, readonly=True, store=True)
-    campaign_id = fields.Many2one('card.campaign', required=True, ondelete='cascade')
-
-    card_element_role = fields.Selection([
-        ('background', 'Background'),
-        ('header', 'Header'),
-        ('subheader', 'Sub-Header'),
-        ('section_1', 'Section 1'),
-        ('subsection_1', 'Sub-Section 1'),
-        ('subsection_2', 'Sub-Section 2'),
-        ('button', 'Button'),
-        ('image_1', 'Image 1'),
-        ('image_2', 'Image 2')
-    ], required=True)
-
-    card_element_image = fields.Image(attachment=False, compute="_compute_value_fields", readonly=False, store=True)
-    card_element_text = fields.Text(compute="_compute_value_fields", readonly=False, store=True)
-    # text displayed in list view
-    card_element_text_value = fields.Char(string="Text", compute="_compute_card_element_text_value")
-    field_path = fields.Char(compute="_compute_value_fields", readonly=False, store=True)
-    text_color = fields.Char(compute="_compute_value_fields", readonly=False, store=True)
-
-    render_type = fields.Selection([('image', 'Image'), ('text', 'User Text')], default='text', required=True)
-    value_type = fields.Selection([('static', 'Manual'), ('field', 'Dynamic')], default='static', required=True)
-
-    _sql_constraints = [('role_uniq', "unique(campaign_id, card_element_role)", "Each campaign should only have one element for each role.")]
-
-    @api.constrains('field_path', 'res_model')
-    def _check_fields(self):
-        skip_security = self.env.su or self.env.user._is_admin()
-        for element in self.filtered(lambda e: e.value_type == 'field'):
-            RenderModel = self.env[element.res_model]
-            field_path = element.field_path
-            if not field_path:
-                raise exceptions.ValidationError(_("field path must be set on %(element_role)s", element_role=element.card_element_role))
-            try:
-                RenderModel.sudo()._find_value_from_field_path(field_path)
-            except (exceptions.UserError, KeyError) as err:
-                raise exceptions.ValidationError(
-                    _('%(model_name)s.%(field_name)s does not seem reachable.',
-                      model_name=RenderModel._name, field_name=field_path)
-                ) from err
-
-            if not skip_security and field_path not in RenderModel._marketing_card_allowed_field_paths():
-                raise exceptions.ValidationError(
-                    _('%(model_name)s.%(field_name)s cannot be used for card campaigns.',
-                      model_name=RenderModel._name, field_name=field_path)
-                )
-
-            path_start, _dummy, last_field = field_path.rpartition('.')
-            # check the last field has a sensible type
-            if element.render_type == 'image' and RenderModel.sudo().mapped(path_start)._fields[last_field].type != 'binary':
-                raise exceptions.ValidationError(
-                    _('%(field_path)s cannot be used as an image value for %(element_role)s',
-                      field_path=field_path, element_role=element.card_element_role)
-                )
-
-    @api.depends('card_element_text', 'field_path')
-    def _compute_card_element_text_value(self):
-        for element in self:
-            field_val = f'[{element.field_path}]' if element.field_path else ''
-            element.card_element_text_value = field_val or element.card_element_text
-
-    @api.depends('render_type', 'value_type')
-    def _compute_value_fields(self):
-        """Reset values irrelevant to the new type of render or value."""
-        for element in self:
-            if element.render_type == 'image':
-                element.text_color = False
-
-            if element.value_type == 'static':
-                element.field_path = False
-
-            if element.value_type == 'field' or element.render_type == 'image':
-                element.card_element_text = False
-
-            if element.value_type == 'field' or element.render_type == 'text':
-                element.card_element_image = False
-
-    def _get_render_value(self, record):
-        """Get the value of the element for a specific record."""
-        self.ensure_one()
-        if record:
-            record.ensure_one()
-        if self.value_type == 'field' and record:
-            # if the value has changed since this was called, we don't know if sudo is allowed
-            if self._origin.field_path != self.field_path:
-                self._check_fields()
-            # this will be called with sudo from the controller, sudo here too for consistency
-            if self.render_type == 'text':
-                return record.sudo()._find_value_from_field_path(self.field_path) or ''
-            return record.sudo().mapped(self.field_path)[0] or ''
-        if self.render_type == 'image':
-            return self.card_element_image or ''
-        if self.render_type == 'text':
-            return self.card_element_text or ''
-        return None
-
-    def _get_placeholder_value(self):
-        """Placeholder to display in preview mode."""
-        self.ensure_one()
-        if self.value_type == 'field':
-            if self.render_type == 'image':
-                return base64.b64encode(self.env['ir.binary']._placeholder())
-            if self.render_type == 'text':
-                return f'[{self.field_path}]'
-        return ''

--- a/addons/marketing_card/models/card_card.py
+++ b/addons/marketing_card/models/card_card.py
@@ -8,14 +8,15 @@ class MarketingCard(models.Model):
     _name = 'card.card'
     _description = 'Marketing Card'
 
+    active = fields.Boolean('Active', default=True)
     campaign_id = fields.Many2one('card.campaign', required=True, ondelete="cascade")
     res_model = fields.Selection(related='campaign_id.res_model')
-    record_ref = fields.Reference(string='Record', selection='_selection_record_ref', compute='_compute_record_ref')
     res_id = fields.Many2oneReference('Record ID', model_field='res_model', required=True)
     image = fields.Image()
+    requires_sync = fields.Boolean(help="Whether the image needs to be updated to match the campaign template.", default=True)
     share_status = fields.Selection([
-        ('shared', 'Shared on Social Networks'),
-        ('visited', 'Share URL Visited'),
+        ('shared', 'Shared'),
+        ('visited', 'Visited'),
     ])
 
     _sql_constraints = [
@@ -24,30 +25,36 @@ class MarketingCard(models.Model):
     ]
 
     @api.depends('res_model', 'res_id')
-    def _compute_record_ref(self):
-        for card in self:
-            card.record_ref = f'{card.res_model},{card.res_id}'
+    def _compute_display_name(self):
+        for model, cards in self.grouped('res_model').items():
+            if not model:
+                cards.display_name = ""
+                continue
+            self.env[model].browse(cards.mapped('res_id')).sudo().fetch(['display_name'])
+            for card in cards:
+                card.display_name = self.env[model].browse(card.res_id).sudo().display_name
+
+    @api.depends('campaign_id')
+    def _compute_res_model(self):
+        """Compute the res_model once and never update it again."""
+        for campaign, cards in self.grouped('campaign_id').items():
+            cards.res_model = campaign.res_model
 
     @api.autovacuum
-    def _gc_card_url_images(self):
-        """Remove images after a day. Social networks are expected to cache the images on their side."""
-        timedelta_days = self.env['ir.config_parameter'].get_param('marketing_card.card_image_cleanup_interval_days', 1)
+    def _gc_card(self):
+        """Remove cards. Social networks are expected to cache the images on their side."""
+        timedelta_days = self.env['ir.config_parameter'].get_param('marketing_card.card_image_cleanup_interval_days', 60)
         if not timedelta_days:
             return
-        self.search([('write_date', '<=', datetime.now() - timedelta(days=timedelta_days))]).image = False
+        self.with_context({"active_test": False}).search([('write_date', '<=', datetime.now() - timedelta(days=timedelta_days))]).unlink()
 
     def _get_card_url(self):
-        return self.campaign_id._get_card_path(self.res_id, 'card.jpg')
+        return self._get_path('card.jpg')
 
     def _get_redirect_url(self):
-        return self.campaign_id._get_card_path(self.res_id, 'redirect')
+        return self._get_path('redirect')
 
-    def _get_or_generate_image(self):
-        # generate if not already stored
-        if not self.image:
-            self.image = self.campaign_id._get_image_b64(self.env[self.res_model].browse(self.res_id))
-        return self.image
-
-    @api.model
-    def _selection_record_ref(self):
-        return self.env['card.campaign']._fields['res_model']._description_selection(self.env)
+    def _get_path(self, suffix):
+        self.ensure_one()
+        card_slug = self.env['ir.http']._slug(self)
+        return f'{self.get_base_url()}/cards/{card_slug}/{suffix}'

--- a/None
+++ b/addons/marketing_card/models/mailing_mailing.py
@@ -0,0 +1,79 @@
+from odoo import _, api, exceptions, fields, models, osv
+
+
+class MassMailing(models.Model):
+    _name = 'mailing.mailing'
+    _inherit = 'mailing.mailing'
+
+    mailing_model_id = fields.Many2one(compute="_compute_mailing_model_id", store=True, readonly=False)
+    card_requires_sync_count = fields.Integer(compute="_compute_card_requires_sync_count")
+    card_campaign_id = fields.Many2one('card.campaign')
+
+    @api.constrains('card_campaign_id', 'mailing_domain', 'mailing_model_id')
+    def _check_mailing_domain(self):
+        for mailing in self:
+            if mailing.card_campaign_id:
+                if mailing.sudo().mailing_model_id.model != mailing.card_campaign_id.res_model:
+                    raise exceptions.ValidationError(_(
+                        "Card Campaign Mailing should target model %(model_name)s",
+                        model_name=self.env['ir.model']._get(mailing.card_campaign_id.res_model).display_name
+                    ))
+
+    @api.depends('card_campaign_id')
+    def _compute_mailing_model_id(self):
+        for mailing in self.filtered('card_campaign_id'):
+            mailing.mailing_model_id = self.env['ir.model']._get_id(mailing.card_campaign_id.res_model)
+
+    @api.depends('card_campaign_id')
+    def _compute_card_requires_sync_count(self):
+        """Check if there's any missing or outdated card."""
+        self.card_requires_sync_count = 0
+        # no point in updating sent mailings
+        card_mailings = self.filtered(lambda mailing: mailing.card_campaign_id and mailing.state == 'draft')
+        for mailing in card_mailings:
+            recipients = self.env[mailing.mailing_model_real].search(self._parse_mailing_domain())
+            out_of_date_count = self.env['card.card'].search_count([
+                ('campaign_id', '=', mailing.card_campaign_id.id),
+                ('res_id', 'in', recipients.ids),
+                ('requires_sync', '=', False)
+            ])
+            mailing.card_requires_sync_count = len(recipients) - out_of_date_count
+
+    def action_put_in_queue(self):
+        """Detect mismatches before scheduling."""
+        for mailing in self.filtered('card_campaign_id'):
+            if mailing.card_requires_sync_count:
+                raise exceptions.UserError(_(
+                    'You should update all the cards for %(mailing)s before scheduling a mailing.',
+                    mailing=mailing.display_name
+                ))
+        super().action_put_in_queue()
+
+    def action_send_mail(self, res_ids=None):
+        for mailing in self.filtered('card_campaign_id'):
+            if mailing.card_requires_sync_count:
+                raise exceptions.UserError(_(
+                    'You should update all the cards for %(mailing)s before scheduling a mailing.',
+                    mailing=mailing.display_name
+                ))
+        return super().action_send_mail(res_ids)
+
+    def action_update_cards(self):
+        """Update the cards in batches, commiting after each batch."""
+        for campaign in self.filtered(lambda mailing: mailing.state == 'draft').card_campaign_id:
+            campaign._update_cards(self._parse_mailing_domain(), auto_commit=True)
+        return {
+            'type': 'ir.actions.act_window',
+            'res_model': 'mailing.mailing',
+            'res_id': self[0].id,
+            'view_mode': 'form',
+            'target': 'current',
+        }
+
+    def _get_recipients_domain(self):
+        """Domain with an additional condition that the card must exist for the records."""
+        domain = super()._get_recipients_domain()
+        if self.card_campaign_id:
+            res_ids = self.env['card.card'].search_fetch([('campaign_id', '=', self.card_campaign_id.id)], ['res_id']).mapped('res_id')
+            domain = osv.expression.AND([domain, [('id', 'in', res_ids)]])
+        return domain

--- a/addons/marketing_card/models/models.py
+++ b/None
@@ -1,34 +0,0 @@
-from odoo import api, models
-
-
-class BaseModel(models.AbstractModel):
-    _inherit = 'base'
-
-    @api.model
-    def _marketing_card_allowed_field_paths(self) -> list[str]:
-        """List of fields allowed to be accessed in template rendering."""
-        allowed_fields_dict = {
-            'res.partner': [
-                'display_name', 'name',
-                'email', 'mobile', 'phone',
-                'country_id', 'country_id.display_name', 'country_id.name',
-                'image_128', 'image_256', 'image_512', 'image_1024',
-            ],
-
-        }
-        allowed_fields_dict['event.track'] = [
-            'display_name', 'name',
-            'event_id', 'event_id.name', 'event_id.display_name',
-            'image', 'partner_id',
-            'location_id', 'location_id.name', 'location_id.display_name',
-            'date', 'date_end',
-            'event_id', 'event_id.name', 'event_id.display_name'
-        ] + [f'partner_id.{partner_path}' for partner_path in allowed_fields_dict['res.partner']]
-        allowed_fields_dict['event.booth'] = [
-            'display_name', 'name',
-            'event_id', 'event_id.name', 'event_id.display_name', 'event_id.address_inline',
-            'event_id.date_begin', 'event_id.date_end', 'event_id.date_tz',
-            'contact_name', 'contact_email', 'contact_phone',
-            'sponsor_name', 'sponsor_email', 'sponsor_mobile', 'sponsor_phone', 'sponsor_subtitle', 'sponsor_image_512',
-        ]
-        return allowed_fields_dict.get(self._name, [])
